bin/migrate_acl_authz.py

```diff
@@ -0,0 +1,191 @@
+    )
+    parser.add_argument(
+        "--arborist-url", dest="arborist", help="URL for the arborist service"
+    )
+    return parser.parse_args()
+
+
+class ACLConverter(object):
+    def __init__(self, sheepdog_db, arborist_url):
+        self.arborist_url = arborist_url.rstrip("/")
+        self.programs = set()
+        self.projects = dict()
+        # map resource paths to tags in arborist so we can save http calls
+        self.arborist_resources = dict()
+
+        engine = create_engine(sheepdog_db, echo=False)
+        try:
+            connection = engine.connect()
+        except OperationalError:
+            raise EnvironmentError(
+                "couldn't connect to sheepdog db using the provided URI"
+            )
+
+        result = connection.execute("SELECT _props->>'name' as name from node_program;")
+        for row in result:
+            self.programs.add(row["name"])
+
+        result = connection.execute("""
+            SELECT
+                project._props->>'name' AS name,
+                program._props->>'name' AS program
+            FROM node_project AS project
+            JOIN edge_projectmemberofprogram AS edge ON edge.src_id = project.node_id
+            JOIN node_program AS program ON edge.dst_id = program.node_id;
+        """)
+        for row in result:
+            self.projects[row["name"]] = row["program"]
+
+        connection.close()
+        return
+
+    def is_program(self, acl_item):
+        return acl_item in self.programs
+
+    def acl_to_authz(self, record):
+        path = None
+        for acl_object in record.acl:
+            acl_item = acl_object.ace
+            if acl_item == "*":
+                path = "/open"
+            elif not path and self.is_program(acl_item):
+                path = "/programs/{}".format(acl_item)
+            else:
+                if acl_item not in self.projects:
+                    raise EnvironmentError(
+                        "program or project {} does not exist".format(acl_item)
+                    )
+                path = "/programs/{}/projects/{}".format(
+                    acl_item, self.projects[acl_item]
+                )
+
+        if not path:
+            logger.error(
+                "couldn't get `authz` for record {} from {}; setting as empty"
+                .format(record.did, record.acl)
+            )
+            return []
+
+        if path not in self.arborist_resources:
+            url = "{}/resource/".format(self.arborist_url)
+            failed = False
+            try:
+                resource = {"path": path}
+                response = requests.post(url, timeout=5, json=resource)
+            except requests.exceptions.Timeout:
+                logger.error(
+                    "couldn't hit arborist to look up resource (timed out): {}".format(url)
+                )
+                failed = True
+            tag = None
+            try:
+                if response.status_code == 409:
```

Comment:
I think you need to check `failed` before this. if it times out then `response` may not exist and this will die

Reply:
good catch :ok_hand: